File: 29Items-Projects/7Algorithms/2GRAPH-ALGORITHMS/java/src/main/java/com/example/graphalgorithms/EdmondsKarp.java
package com.example.graphalgorithms;

import java.util.*;

public class EdmondsKarp {
    public static int maxFlow(int[][] capacity, int source, int sink) {
        validate(capacity, source, sink);
        int[][] residual = copy(capacity);
        int maxFlow = 0;
        int[] parent = new int[capacity.length];
        while (bfs(residual, source, sink, parent)) {
            int pathFlow = Integer.MAX_VALUE;
            for (int v = sink; v != source; v = parent[v]) {
                pathFlow = Math.min(pathFlow, residual[parent[v]][v]);
            }
            for (int v = sink; v != source; v = parent[v]) {
                residual[parent[v]][v] -= pathFlow;
                residual[v][parent[v]] += pathFlow;
            }
            maxFlow += pathFlow;
        }
        return maxFlow;
    }

    private static boolean bfs(int[][] residual, int source, int sink, int[] parent) {
        Arrays.fill(parent, -1);
        boolean[] visited = new boolean[residual.length];
        ArrayDeque<Integer> queue = new ArrayDeque<>();
        queue.add(source);
        visited[source] = true;
        while (!queue.isEmpty()) {
            int current = queue.remove();
            for (int next = 0; next < residual.length; next++) {
                if (!visited[next] && residual[current][next] > 0) {
                    parent[next] = current;
                    if (next == sink) {
                        return true;
                    }
                    visited[next] = true;
                    queue.add(next);
                }
            }
        }
        return false;
    }

    private static int[][] copy(int[][] matrix) {
        int[][] result = new int[matrix.length][];
        for (int i = 0; i < matrix.length; i++) {
            result[i] = Arrays.copyOf(matrix[i], matrix[i].length);
        }
        return result;
    }

    private static void validate(int[][] capacity, int source, int sink) {
        if (capacity.length == 0) {
            throw new IllegalArgumentException("capacity matrix is required");
        }
        if (source == sink || source < 0 || sink < 0 || source >= capacity.length || sink >= capacity.length) {
            throw new IllegalArgumentException("source and sink must be valid distinct indexes");
        }
        for (int[] row : capacity) {
            if (row.length != capacity.length) {
                throw new IllegalArgumentException("capacity matrix must be square");
            }
            for (int value : row) {
                if (value < 0) {
                    throw new IllegalArgumentException("capacities must be non-negative");
                }
            }
        }
    }

    public static void main(String[] args) {
        int[][] capacity = {
            {0, 16, 13, 0, 0, 0},
            {0, 0, 10, 12, 0, 0},
            {0, 4, 0, 0, 14, 0},
            {0, 0, 9, 0, 0, 20},
            {0, 0, 0, 7, 0, 4},
            {0, 0, 0, 0, 0, 0}
        };
        System.out.println("Edmonds-Karp max flow");
        System.out.println(maxFlow(capacity, 0, 5));
    }
}

File: 29Items-Projects/7Algorithms/2GRAPH-ALGORITHMS/python/graph_algorithms/edmonds_karp.py
from __future__ import annotations

from collections import deque


def _bfs(residual: list[list[int]], source: int, sink: int, parent: list[int]) -> bool:
    visited = {source}
    queue = deque([source])
    while queue:
        current = queue.popleft()
        for vertex, capacity in enumerate(residual[current]):
            if vertex not in visited and capacity > 0:
                parent[vertex] = current
                if vertex == sink:
                    return True
                visited.add(vertex)
                queue.append(vertex)
    return False


def edmonds_karp(capacity: list[list[int]], source: int, sink: int) -> int:
    _validate_capacity(capacity, source, sink)
    residual = [row[:] for row in capacity]
    max_flow = 0

    while True:
        parent = [-1] * len(capacity)
        if not _bfs(residual, source, sink, parent):
            return max_flow
        path_flow = 10**9
        vertex = sink
        while vertex != source:
            previous = parent[vertex]
            path_flow = min(path_flow, residual[previous][vertex])
            vertex = previous
        vertex = sink
        while vertex != source:
            previous = parent[vertex]
            residual[previous][vertex] -= path_flow
            residual[vertex][previous] += path_flow
            vertex = previous
        max_flow += path_flow


def _validate_capacity(capacity: list[list[int]], source: int, sink: int) -> None:
    if not capacity:
        raise ValueError("capacity matrix is required")
    if any(len(row) != len(capacity) for row in capacity):
        raise ValueError("capacity matrix must be square")
    if source == sink:
        raise ValueError("source and sink must be different")
    if source < 0 or sink < 0 or source >= len(capacity) or sink >= len(capacity):
        raise ValueError("source and sink must be valid vertex indexes")
    if any(value < 0 for row in capacity for value in row):
        raise ValueError("capacities must be non-negative")


if __name__ == "__main__":
    capacities = [
        [0, 16, 13, 0, 0, 0],
        [0, 0, 10, 12, 0, 0],
        [0, 4, 0, 0, 14, 0],
        [0, 0, 9, 0, 0, 20],
        [0, 0, 0, 7, 0, 4],
        [0, 0, 0, 0, 0, 0],
    ]
    print("Edmonds-Karp max flow")
    print(edmonds_karp(capacities, 0, 5))
